Automattic/wp-calypso
# Modularize Reader state

## 📢Let the grand modularization of state of 2020... commence! 📢

Ahem.

This is the first of what are likely to be many PRs attempting to introduce some modularization of state to Calypso. This one deals with Reader, specifically. As the first of the PRs, it also introduces the mechanism through which modularization will happen.

 See p4TIVU-9lM-p2

### How state modularization works

We don't make any changes to reducers, selectors, or actions themselves for this to work, nor do we change the Redux architecture in any way. The approach we take builds on top of Redux by adding a new, developer-enforced requirement: creating static dependencies between actions/selectors and the reducers that control the parts of state they're interested in.

For this to work, reducers first need to be able to register themselves. That's done with the addition of a global Redux store handler, which exposes a registration method for reducers. It coordinates registration with store creation through the use of a synchronous queue.

Once a reducer is able to register itself, we can add the registration as a side effect to its main module. Other modules that make use of that reducer's corresponding state, such as action and selector modules, can then simply perform a naked import of the reducer module, to ensure that it's loaded (and thus registered) in time for their methods to be used.

This way, we achieve modularization through the dependency graph, and get code splitting for free through the regular build process. The code moves from upfront chunks that get loaded on every entry point to only the relevant chunks, removing the need for every entry point to load all of state, and thus reducing the amount of JS the end user needs to load, parse, and execute on any given route.

### Tradeoff for consideration

Redux state is a tree, and we could arguably set modularization at any level in that tree. Finer granularity gets us more potential savings, at the cost of more development overhead: it's easier to add a single import for the reader reducer than it is to add one for reader conversations, another for reader feeds, etc.

In this PR, I opted for a very coarse level of granularity, by splitting at the top-level key. I expect this to offer a good tradeoff in terms of getting some significant gains while keeping things simple, but I'd like to hear your thoughts.

### Changes proposed in this Pull Request

* Add global Redux store handler
* Set global Redux store on boot (for the `entry-main` and `entry-login` entry points)
* Remove Reader reducer instantiation from `entry-main` boot
* Add naked imports of Reader reducer to all Reader-related actions and selectors
* Minor lint fixes

I split this PR in three separate commits to facilitate reviewing the different aspects separately.

### Testing instructions

* Install [Redux DevTools](https://chrome.google.com/webstore/detail/redux-devtools/lmhkpmbekcpmknklioeibfkpmmfibljd?hl=en).
* Open DevTools and switch to the Redux tab.
* Open `/home`.
* Click `State` on the top right to ensure you see all state, and not just the diff for the last action: 
![image](https://user-images.githubusercontent.com/409615/73774406-d3e87d80-477b-11ea-9f59-3e42cc00101f.png)
* Analyze the tree and note that there's no `reader` key, even if you expand the list of keys. This indicates that the Reader state hasn't been loaded yet.
* Click `Reader` on the masterbar.
* Verify that a `reader` key is added with the Reader state. This indicates that the Reader state has now been loaded.
* Verify that Reader works normally. This indicates that I didn't mess up.
